### PR TITLE
[eas-cli] replace got with node-fetch

### LINKED ISSUE #32: Use node-fetch for all network requests: node-fetch and got both installed in eas-cli
should settle on one for networking. iirc we've tried to stick to axios or got in expo-cli.

### PR DESCRIPTION
**Pull request by wkozyra95**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

Closes #32 

I plan to add proxy support in the next PR, so I decided to do this first to not duplicate the work

# How

replace got with node-fetch

# Test Plan

run tests
run `eas build` , `eas submit` and `eas update`

### CODE REVIEW

File: packages/eas-cli/src/api.ts [hunk 1]
```diff
@@ -1,42 +1,91 @@
...
+    await this.handleApiErrorAsync(response);
+    return await response.json();
+  }
+
+  private async handleApiErrorAsync(response: Response): Promise<void> {
+    if (response.status >= 400) {
+      let result: { [key: string]: any };
+      try {
+        result = await response.json();
+      } catch (err) {
```
Comment (kbrandwijk) [inline packages/eas-cli/src/api.ts:66]: Untyped err? `response.json()` will raise a `SyntaxError`, so err type is known.

**Review by kbrandwijk — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/api.ts [hunk 2]
```diff
@@ -1,42 +1,91 @@
...
+}
+
+class ApiV2 {
+  public async putAsync(path: string, options: RequestOptions): Promise<any> {
+    return await this.requestAsync(path, { method: 'PUT', body: JSON.stringify(options.body) });
+  }
+
+  public async postAsync(path: string, options: RequestOptions): Promise<any> {
+    return await this.requestAsync(path, { method: 'POST', body: JSON.stringify(options.body) });
+  }
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/api.ts:39]: What other options? RequestOptions has only one field

Considered just passing body there, but this is more consistent with how it worked with got.

File: packages/eas-cli/src/api.ts (hunk 1 above)
Comment (wkozyra95) [reply packages/eas-cli/src/api.ts:66]:
```suggestion
      } catch {
```

File: packages/eas-cli/src/api.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kbrandwijk) [reply packages/eas-cli/src/api.ts:39]: https://github.com/node-fetch/node-fetch#options

[the remark this answers is not in the document]
Comment (kbrandwijk) [reply packages/eas-cli/src/api.ts:39]: Oh, I now see that RequestOptions is actually not the 'regular' RequestOptions, but a custom one with the same name. That's confusing...

File: packages/eas-cli/src/utils/__tests__/code-signing-test.ts [hunk 3]
```diff
@@ -97,10 +97,10 @@ describe(parseMultipartMixedResponseAsync, () => {
     const form = generateMultipartBody(JSON.stringify({ hello: 'world' }));
 
     const parts = await parseMultipartMixedResponseAsync({
-      rawBody: form.getBuffer(),
-      headers: {
+      arrayBuffer: () => Promise.resolve(form.getBuffer()),
```
Comment (EvanBacon) [inline packages/eas-cli/src/utils/__tests__/code-signing-test.ts:100]:
```suggestion
      arrayBuffer: async () => form.getBuffer(),
```
I think we generally opt towards this syntax in expo (based on recent PR feedback I've been getting). I have no preference.

File: packages/eas-cli/src/fetch.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import fetch, { RequestInfo, RequestInit, Response } from 'node-fetch';
+
+export * from 'node-fetch';
+
+export default async function (url: RequestInfo, init?: RequestInit): Promise<Response> {
+  const response = await fetch(url, init);
+  if (response.status >= 400) {
+    const error = new Error(`Request failed: ${response.status} (${response.statusText})`);
+    (error as any).response = response;
+    throw error;
```
Comment (EvanBacon) [inline packages/eas-cli/src/fetch.ts:10]: Probably makes sense to create a new Error subclass that has the response property.

**Review by EvanBacon — COMMENTED** [review]
This mostly looks fine. I think one thing worth noting is that we'll eventually want a unified package for interacting with the Expo API in Expo Go, expo.dev, Expo CLI, EAS CLI. 

Moving to node-fetch is a great first step regardless of how we do this.

File: packages/eas-cli/src/api.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/api.ts:39]: I could rename it but type provided by node-fetch is not named RequestOptions, so I don't see what name would be better. Any suggestions?

[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/api.ts:39]: to clarify type defined in typescript is called `RequestInit` and node-fetch defines it as `Options` in documentation, but it's not something the developer would from code completion or anything like this.

[the remark this answers is not in the document]
Comment (kbrandwijk) [reply packages/eas-cli/src/api.ts:39]: I thought I saw node-fetch actually had it defined as `RequestOptions` when I checked last week. `http` does have a `RequestOptions`, which is also used internally in node-fetch, but not exposed as such.

**Review by kbrandwijk — APPROVED** [review]

File: packages/eas-cli/src/api.ts [hunk 5]
```diff
@@ -1,42 +1,96 @@
...
+    }
+    const sessionSecret = getSessionSecret();
+    if (sessionSecret) {
+      return { 'expo-session': sessionSecret };
+    }
+    return {};
+  }
+}
+
+export const api = new ApiV2();
```
Comment (wschurman) [inline packages/eas-cli/src/api.ts:93]: optional: alternatively this class could be all static methods since it doesn't need state

**Review by wschurman — APPROVED** [review]

**Pull request merged**